File: Algorithms/leetcode/Ex_0376_Wiggle_Subsequence/Soln_2_Dp_Space_O1.cpp
#include <vector>

using std::vector;
using std::max;

class Solution_2 {
public:
    int wiggleMaxLength(vector<int>& nums) {
        int size = nums.size();
        
        if (size == 0) {
            return 0;
        }

        int up = 1;
        int down = 1;

        // 开始填表
        for (int i = 1; i < size; i++) {
            if (nums[i] > nums[i-1]) {
                up = down + 1;
            } else if (nums[i] < nums[i-1]) {
                down = up + 1;
            }
        }

        return max(up, down);
    }
};

// 参考题解（非常推荐，图解很清楚）：https://leetcode-cn.com/problems/wiggle-subsequence/solution/tan-xin-si-lu-qing-xi-er-zheng-que-de-ti-jie-by-lg/

File: Algorithms/leetcode/Ex_0674_Longest_Continuous_Increasing_Subsequence/Soln_2_Dp_Space_O1.java
package Algorithms.leetcode.Ex_0674_Longest_Continuous_Increasing_Subsequence;

class Solution_2 {
    public int findLengthOfLCIS(int[] nums) {
        int length = nums.length;

        // 特判
        if (length <= 1) {
            return length;
        }

        // curLen 记录当前记录的最长连续递增序列的长度
        int curLen = 1;

        int res = 1;    // 记录全局的最长连续递增序列的长度
        

        for (int i = 1; i < length; i++) {
            if (nums[i] > nums[i-1]) {
                curLen ++;
                res = Math.max(res, curLen);
            } else {
                // 重置 curLen，从当前 nums[i] 重新开始计算递增序列
                curLen = 1;
            }
        }

        return res;

    }
}

// 参考题解（非常推荐）：https://leetcode-cn.com/problems/number-of-longest-increasing-subsequence/solution/dong-tai-gui-hua-jie-zui-chang-zi-xu-lie-zi-chua-4/
// 建议和参考题解中的其他相关题目一起做，效果会更好

